bokeh/core/validation/check.py

```diff
@@ -117,17 +126,40 @@ def check_integrity(models):
         for func in validators:
             messages[func.validator_type].extend(func())
 
-    for msg in sorted(messages['error']):
+    return messages
+
+def process_validation_issues(issues):
+    ''' Log warning and error messages for a dictionary containing warnings and error messages.
+
+    Args:
+        issues (dict(error=[], warning=[])) : A dictionary of all warning and error messages
+
+    Returns:
+        None
+
+    This function will emit log warning and error messages for all error or
+    warning conditions in the dictionary. For example, a dictionary
+    containing a warning for empty layout will trigger a warning:
+
+    .. code-block:: python
+
+        >>> process_validation_issues(validations)
+        W-1002 (EMPTY_LAYOUT): Layout has no children: Row(id='2404a029-c69b-4e30-9b7d-4b7b6cdaad5b', ...)
+
+    '''
+    for msg in sorted(issues['error']):
         log.error("E-%d (%s): %s: %s" % msg)
 
-    for msg in sorted(messages['warning']):
+    for msg in sorted(issues['warning']):
         code, name, desc, obj = msg
         if code not in __silencers__:
             log.warning("W-%d (%s): %s: %s" % msg)
-
-    # This will be turned on in a future release
-    # if len(messages['error']) or (len(messages['warning']) and settings.strict()):
-    #     raise RuntimeError("Errors encountered during validation (see log output)")
+    if settings.validation_level() == "errors":
+        if len(issues['error']):
+            raise RuntimeError("Errors encountered during validation (see log output)")
+    elif settings.validation_level() == "all":
+        if len(issues['error']) or len(issues['warning']):
+            raise RuntimeError("Errors encountered during validation (see log output)")
```

Comment:
We should not just say "see log output", the Runtime error should itself should be formatted to include all the warnings and/or errors as appropriate.

Reply:
I might suggest a code structure more like this:
```
# filter things first so it's not an issue later
errors = issues['errors']
warnings = [item for item in issues['warnings'] if item[0] not in __silencers__

# collect all the message that will get logged as they are logged
warning_messages = [] 
for code, name, desc, obj in sorted(warnings):
    msg = f"W-{code} ({name}): {desc}: {obj}"
    warning_messages.append(msg)
    log.warning(msg)

# do similar with errors

# raise exceptions, using warning_messages and errror_messages
```

Reply:
Also as a remindedr we always use f-strings, and do not use old-style string formatting anywhere.